sdks/java/extensions/sbe/src/main/java/org/apache/beam/sdk/extensions/sbe/SbeSchema.java

```diff
@@ -0,0 +1,169 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+import uk.co.real_logic.sbe.ir.Ir;
+
+/**
+ * Represents an SBE schema that can be translated to a Beam {@link Schema} and {@link
+ * PayloadSerializerProvider}.
+ *
+ * <p>The schema represents a single SBE message. If the XML schema contains more than one message,
+ * then a new instance must be created for each message that the pipeline will work with.
+ *
+ * <p>The currently supported ways of generating a schema are:
+ *
+ * <ul>
+ *   <li>Through an intermediate representation ({@link Ir}).
+ * </ul>
+ *
+ * <h3>Intermediate Representation</h3>
+ *
+ * <p>An {@link Ir} allows for a reflection-less way of getting a very accurate representation of
+ * the SBE schema, since it is a tokenized form of the original XML schema. To help deal with some
+ * ambiguities, such as which message to base the schema around, passing {@link IrOptions} is
+ * required. See the Javadoc for the options for more details.
+ */
+@Experimental(Kind.SCHEMAS)
+public final class SbeSchema implements Serializable {
+  private static final long serialVersionUID = 1L;
+
+  private final @Nullable SerializableIr ir;
+  private final @Nullable IrOptions irOptions;
+  private final ImmutableList<SbeField> sbeFields;
+
+  private SbeSchema(
+      @Nullable SerializableIr ir,
+      @Nullable IrOptions irOptions,
+      ImmutableList<SbeField> sbeFields) {
+    this.ir = ir;
+    this.irOptions = irOptions;
+    this.sbeFields = sbeFields;
+  }
+
+  /**
+   * Creates a new {@link SbeSchema} from the given intermediate representation.
+   *
+   * <p>This makes no guarantees about the state of the returned instance. That is, it may or may
+   * not have the generated SBE schema representation, and it may or may not have translated the SBE
+   * schema into a Beam schema.
+   *
+   * @param ir the intermediate representation of the SBE schema. Modifications to the passed-in
+   *     value will not be reflected in the returned instance.
+   * @param irOptions options for configuring how to deal with cases where the desired behavior is
+   *     ambiguous.
+   * @return a new {@link SbeSchema} instance
+   */
+  public static SbeSchema fromIr(Ir ir, IrOptions irOptions) {
+    ImmutableList<SbeField> sbeFields = IrFieldGenerator.generateFields(ir, irOptions);
+
+    Ir copy =
+        new Ir(
+            ir.packageName(),
+            ir.namespaceName(),
+            ir.id(),
+            ir.version(),
+            ir.description(),
+            ir.semanticVersion(),
+            ir.byteOrder(),
+            ImmutableList.copyOf(ir.headerStructure().tokens()));
+
+    return new SbeSchema(SerializableIr.fromIr(copy), irOptions, sbeFields);
+  }
+
+  @VisibleForTesting
+  @Nullable
+  Ir getIr() {
+    return ir == null ? null : ir.ir();
+  }
+
+  @VisibleForTesting
+  @Nullable
+  IrOptions getIrOptions() {
+    return irOptions;
+  }
+
+  @VisibleForTesting
```

Comment:
`@VisibleForTesting` is a bit of an anti-pattern in my opinion. We should be testing through the public APIs as much as possible. Currently, it looks like this class doesn't actually _have_ a public API, you can create an instance through a public function, but all the accessors indicate they're exposed just for testing. Should some of these accessors be public?

Reply:
My idea here was to make sure that modifications to the passed-in value couldn't also modify the `SbeSchema` instance, though in that case, it would probably make more sense to read the field directly in tests.

However, there's no real problem making these public, and I could see some use in making that info public (e.g. logging if something goes wrong), so I'll just make the methods public.